### ISSUE TITLE
[FEA] Parallelize shuffle compress/decompress with opportunistic idle task threads

### ISSUE THREAD

**Issue by jlowe** [NVIDIA/spark-rapids#5039]
**Is your feature request related to a problem? Please describe.**
Nsight traces show that many queries have cases where one or more threads are blocked trying to acquire the GPU while one or more other threads are compressing output shuffle data.  It would be nice to make use of CPU threads that are blocked on GPU acquisition to help out on tasks that are compressing their shuffle outputs.

**Describe the solution you'd like**
Here's one idea to try to make use of the idle task threads.  The idea here is to try to honor the executor's configured CPU resources by making sure we're not creating active background threads corresponding to active task threads.  Background threads would only be allowed to be active while the task thread is sleeping on the GPU semaphore (with the caveat of overlap is allowed as the background thread finishes its current task).  I'm thinking something along these high-level lines:

- Create a thread pool that can have a maximum number of threads as the number of executor task threads minus the configured GPU task concurrency (i.e.: the maximum number of task threads that could block on the GPU semaphore).  However the thread pool starts out with zero eligible threads to perform work.
- Change `GpuSemaphore` such that if a task thread is going to block on the GPU semaphore then we activate/wakeup one of the threads in the thread pool created above by increasing the number of threads allowed in the thread pool.  That thread is now eligible to "help out" on CPU activities.
- When the blocked task thread ultimately acquires the GPU semaphore, it lowers the thread count in the thread pool, effectively removing that thread from participating in CPU-assist activities after it finishes its current work (if any).

The shuffle write could create a work queue that is shared between the current task thread and the "CPU assist" thread pool workers.  Any activated CPU-assist pool threads could help serialize and compress shuffle output data in parallel with the task thread.  This probably requires a custom shuffle manager, but we may be able to wrap the existing shuffle manager in some way to keep it as compatible as possible with the cluster's shuffle configuration.

A similar setup may be useful for shuffle read, where the CPU-assist thread pool can help decompress/deserialize partitions in parallel to help on the shuffle read side.

**Comment by abellina** [comment]
I just wanted to add that one first step is to try and create a quick and dirty PoC to make sure we understand what the potential gains might be. From the looks of it, many of the queries we've seen that are 10s+ in time for the 3TB dataset can spend more time reading and writing (specifically compressing/decompressing) shuffle blocks than actual compute.

**Comment by abellina** [comment]
I think for now I can do the PoC and show whether it improves things or not, and then we can decide how to do it right, and who works on that.

**Comment by abellina** [comment]
I added a separate ticket specific for the write path: https://github.com/NVIDIA/spark-rapids/issues/6060

**Comment by abellina** [comment]
For the read path: https://github.com/NVIDIA/spark-rapids/issues/6061

**Comment by abellina** [comment]
As part of this issue, we need to figure out how to make the config/thread pool added https://github.com/NVIDIA/spark-rapids/issues/6060 automatically size itself and be as turnkey as possible (re comment: https://github.com/NVIDIA/spark-rapids/pull/6052#discussion_r927806550)

**Comment by abellina** [comment]
Continuing part of this in 22.12. Specifically, the smarts around how these thread pools should be sized and potentially working with the semaphore. What we have now is statically configured, and we want to go away from users manually tuning configs.
